components/script/dom/node.rs

```diff
@@ -1858,6 +1858,14 @@ impl NodeMethods for Node {
         }
     }
 
+    // https://dom.spec.whatwg.org/#dom-node-rootnode
+    fn RootNode(&self) -> Root<Node> {
+        match self.GetParentNode() {
+            Some(parent) => parent.RootNode(),
```

Comment:
In this case it seems easy enough to make this an explicit loop roughly as follows (since Rust [doesn't actually have tail-call optimization](https://github.com/rust-lang/rust/issues/217)):
```rust
let current = &self;
while let Some(ref parent) = current.GetParentNode() {
    current = parent;
}
Root::from_ref(current)
```